airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py

```diff
@@ -66,14 +66,15 @@ def on_partition_generation_completed(self, sentinel: PartitionGenerationComplet
         """
         stream_name = sentinel.stream.name
         self._streams_currently_generating_partitions.remove(sentinel.stream.name)
-        ret = []
         # It is possible for the stream to already be done if no partitions were generated
         # If the partition generation process was completed and there are no partitions left to process, the stream is done
         if self._is_stream_done(stream_name) or len(self._streams_to_running_partitions[stream_name]) == 0:
-            ret.append(self._on_stream_is_done(stream_name))
+            yield self._on_stream_is_done(stream_name)
+            if not sentinel.has_generated_partition:
+                sentinel.stream.cursor.emit_state_given_no_partitions_generated()
+                yield from self._message_repository.consume_queue()
```

Comment:
I'm probably being slow, but what is the reasoning behind emitting the state before consuming from the queue? If the previous lines have generated messages that go into the repository it feels like they should be emitted first.

Reply:
Emitting the state only push it to the message repository queue. The owner of actually ordering/printing the records decide when to emit these. 

I don't really like this logic because as a user, I would assume that `emit state` would actually send it. I don't remember why we did it like that. That being said, it has advantages: we can send messages to the message repository in a multi threaded way and emit them only on the main thread which allows us to avoid issues regarding messages order

Reply:
Ah right, that makes sense!

Reply:
> we can send messages to the message repository in a multi threaded way and emit them only on the main thread which allows us to avoid issues regarding messages order

+1. super important to avoid emitting state messages before the records